src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Tree/Dependencies/DependenciesTreeBuilder.cs

```diff
@@ -0,0 +1,594 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using Microsoft.VisualStudio.Composition;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.ProjectSystem.References;
+using Microsoft.VisualStudio.ProjectSystem.Tree.Dependencies.Snapshot;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Tree.Dependencies;
+
+/// <summary>
+/// Populates and updates an <see cref="IProjectTree"/> from a <see cref="DependenciesSnapshot"/>.
+/// </summary>
+/// <remarks>
+/// This class's primary entry point is <see cref="BuildTreeAsync"/>.
+/// </remarks>
+[Export]
+internal sealed class DependenciesTreeBuilder
+{
+    private static ImmutableDictionary<ProjectConfigurationSlice, ProjectTreeFlags> s_flagsByConfigurationSlice = ImmutableDictionary<ProjectConfigurationSlice, ProjectTreeFlags>.Empty;
+
+    private readonly UnconfiguredProject _unconfiguredProject;
+
+#if false
+    private DependenciesSnapshot? _lastSnapshot;
+#endif
+
+    /// <summary>
+    /// <see cref="IProjectTreePropertiesProvider"/> imports that apply to the references tree.
+    /// </summary>
+    [ImportMany(ReferencesProjectTreeCustomizablePropertyValues.ContractName)]
+    private readonly OrderPrecedenceImportCollection<IProjectTreePropertiesProvider> _projectTreePropertiesProviders;
+
+    [Import]
+    internal IProjectTreeOperations TreeConstruction { get; set; } = null!;
+
+    [ImportingConstructor]
+    public DependenciesTreeBuilder(UnconfiguredProject unconfiguredProject)
+    {
+        _unconfiguredProject = unconfiguredProject;
+
+        _projectTreePropertiesProviders = new OrderPrecedenceImportCollection<IProjectTreePropertiesProvider>(
+            ImportOrderPrecedenceComparer.PreferenceOrder.PreferredComesLast,
+            projectCapabilityCheckProvider: unconfiguredProject);
+    }
+
+    /// <summary>
+    /// Builds the "Dependencies" node <see cref="IProjectTree"/> for the given <paramref name="snapshot"/> based on the previous <paramref name="dependenciesNode"/>.
+    /// </summary>
+    /// <param name="dependenciesNode">The previous dependencies tree, to which the updated <paramref name="snapshot"/> should be applied.</param>
+    /// <param name="snapshot">The current dependencies snapshot to apply to the tree.</param>
+    /// <param name="cancellationToken">Supports cancellation of this operation.</param>
+    /// <returns>An updated "Dependencies" node.</returns>
+    internal async Task<IProjectTree> BuildTreeAsync(
```

Comment:
> all of these functions could become private methods

That wouldn't change the size of the code. To my mind it would increase the complexity, because having N top-level functions has more unbounded complexity than scoping each to a smaller number of potential call sites.

We have this discussion from time to time. I really like scoping things to as small a scope as possible, and local functions (like local variables) let you do that.

At the end of the day it's mostly a matter of preference.